Added Mollie terminal support for Odoo POS and Kiosk

pga-odoo opened the pull request:
Added 2 new module for mollie terminal payments for POS, here are the details,
- pos_mollie module adds Mollie terminal support for regular POS
- pos_self_order_mollie adds Mollie terminal support for kiosk POS

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/147442).

vlst-odoo commented on addons/pos_self_order_mollie/controllers/main.py:
why would you have 2 webhooks for the same thing? You can override `_mollie_process_webhook` instead of create a whole new controller

pga-odoo replied:
Mollie webhook only send payment ID. So we do not know in prior the webhook called for pos payment or kiosk payment.  We need to call mollie endpoint to know payment related data. So if we manage it in single controller then pos payment will make 2 mollie api call.  So to avoid that we need to submit data to a common method and override that method instead of controller. So to avoid all that i have added have 2 webhooks.
 
 If you want i can do common controller with type as path `/pos_mollie/webhook/<str:payment_type>/<int:payment_method_id>`
 And i will submit order type from the POS so we can remove `_get_mollie_webhook_url` method completly 
 What do you think ?

vlst-odoo replied:
This sounds good. We can go for this approach

vlst-odoo commented on addons/pos_self_order_mollie/models/pos_payment_method.py:
```diff
@@ -0,0 +1,59 @@
+            result = self.mollie_payment_request(payment_data)
+            return result and result.get('status') == 'open'
+
+    def _get_mollie_webhook_url(self, data):
+        webhook_url = super()._get_mollie_webhook_url(data)
+        if data.get('order_type') == 'kiosk':
+            base_url = self.get_base_url()
+            webhook_url = urls.url_join(base_url, f'/pos_mollie_kiosk/webhook/{self.id}')
+        return webhook_url
```
this can be removed after you remove the kiosk webhook

pga-odoo replied:
Please see details in previous comment

vlst-odoo commented on addons/pos_mollie/static/src/app/payment_mollie.js:
```diff
@@ -0,0 +1,176 @@
+    async send_payment_cancel(order, cid) {
+        const { confirmed } = await this.env.services.popup.add(ConfirmPopup, {
+            title: _t('Cancel mollie payment'),
+            body: _t('First cancel transaction on POS device. Only use force cancel if that fails'),
```
this doesn't feel extremely intuitive... The user will be tempted to click on the confirmation button, which is not what you want in this case

pga-odoo replied:
Yes you are right.

I have removed message instead, i have cancelled payment and show the message `Mollie payments also need manual cancellation on the payment terminal.` 

Is it ok ?

vlst-odoo replied:
Nice!

vlst-odoo commented on addons/pos_mollie/static/src/app/payment_mollie.js:
```diff
@@ -0,0 +1,176 @@
+        }
+    }
+
+    _handle_odoo_connection_failure(data = {}) {
+        const line = this.get_pending_mollie_line();
+        if (line) {
+            line.set_payment_status("retry");
+        }
+        this._show_error(
+            _t("Could not connect to the Odoo server, please check your internet connection and try again.")
+        );
+        return Promise.reject(data);
+    }
```
this is the same function as in `pos_adyen`. Can you find a way to not duplicate it? Perhaps lift it to the `PaymentInterface` class

pga-odoo replied:
This method is available in several other 3 pos payment modules, i think better do it in another task.

What do you think ?

vlst-odoo replied:
You are right. We can merge like this and we'll handle that later

vlst-odoo commented on addons/pos_mollie/static/src/app/payment_mollie.js:
```diff
@@ -0,0 +1,176 @@
+
+    }
+
+    waitForPaymentConfirmation() {
+        return new Promise((resolve) => {
+            this.paymentLineResolvers[this.get_pending_mollie_line().cid] = resolve;
+        });
+    }
```
same idea, try to not duplicate this

vlst-odoo commented on addons/pos_mollie/static/src/app/payment_mollie.js:
```diff
@@ -0,0 +1,176 @@
+        return this.pos.getPendingPaymentLine("mollie");
+    }
+
+    _show_error(msg, title) {
+        if (!title) {
+            title = _t("Mollie Error");
+        }
+        this.env.services.popup.add(ErrorPopup, {
+            title: title,
+            body: msg,
+        });
+    }
+}
```
lift this to `PaymentInterface` too

pga-odoo replied:
This method is available in 2 other pos payment modules, i think better do it in another task.

What do you think ?

vlst-odoo commented on addons/pos_self_order_mollie/static/description/icon.png:
you don't need this image 2 times. You can just use the one from `pos_mollie` in kiosk too

pga-odoo replied:
ok, i have replaced it with the same image as `pos_mollie`

vlst-odoo commented on addons/pos_mollie/static/src/app/payment_mollie.js:
```diff
@@ -0,0 +1,176 @@
+    /**
+     * Call odoo backend to create payment request on mollie.
+     */
+    _prepare_payment_data(cid) {
+        const order = this.pos.get_order();
+        const line = order.paymentlines.find((paymentLine) => paymentLine.cid === cid);
+        return {
+            'description': order.name,
+            'pos_reference': order.uid,
+            'currency': this.pos.currency.name,
+            'amount': line.amount,
+            'session_id': this.pos.pos_session.id,
+            'config_id': this.pos.pos_session.config_id[0],
+        }
+    }
```
here you prepare the data a bit and then on the serve you prepare it some more. Perhaps from js you should just what paymentline you want to pay for and perform all the "preparation" on the server

pga-odoo replied:
Hello,

Some data like order name, uid , session_id etc are not available on the server side and we need it in webhook processing. 

I removed `config_id` as it was needed.

pga-odoo commented on addons/pos_mollie/static/src/app/pos_bus.js:
Moved file `overrides/models folder`

I just followed other pos payment modules. Both `pos_adyen`, `pos_viva_wallet`  have it under app folder.

vlst-odoo replied:
i agree with you. It's a mistake in those places

pga-odoo commented:
Hello @vlst-odoo

I have updated some code as per your feedback. For some i have added comments.

Can you please check? Let me know if something is not clear.

vlst-odoo commented:
> Hello @vlst-odoo
> 
> I have updated some code as per your feedback. For some i have added comments.
> 
> Can you please check? Let me know if something is not clear.

Hello, nice work!

I agree with your suggestion for how to unify the controllers and i'll take a final look after you do that

vlst-odoo commented on addons/pos_self_order_mollie/controllers/main.py:
i think that you can get rid of this entire file and perhaps just override the `_mollie_process_webhook` method to take into account the distinction between pos and kiosk. Also, if this helps, just by knowing the pos_reference of an order you can know if it is a kiosk or pos order

pga-odoo replied:
Good Idea :+1:  I have removed `main.py` as per your suggestion and manged kiosk with overriding `_mollie_process_webhook`. 

Mollie webhook submits only transaction ID of mollie payment. We are not getting `pos_reference`. We make API call to access all other metadata.

vlst-odoo commented on addons/pos_self_order_mollie/static/description/icon.png:
this image can be removed. It already exists in `pos_mollie`

pga-odoo replied:
removed icon

pga-odoo commented:
Hello @vlst-odoo 

I made changes according to your feedback.

Let me know if you need something else.

pga-odoo commented:
Hello @vlst-odoo 

Do you have any updates? Please inform me if there are any changes required on my end.

I've also prepared the SAAS-17.1 module for forward porting.

vlst-odoo commented:
> Hello @vlst-odoo
> 
> I made changes according to your feedback.
> 
> Let me know if you need something else.

Would you be able to address some of these concerns ? For example the `csrf` issue ?
![image](https://github.com/odoo/odoo/assets/118446179/56e4f72a-adca-425a-85c1-65f67e21339c)

The bus communication can be done with the `notify` method in the fwports, but for now you will need the security override for it.

pga-odoo commented:
@vlst-odoo 

csrf needs to be false in the as it will be called from mollie so they will not have csrf token. (all of the webhooks in odoo is having csrf false see) 

- [pos_adyen](https://github.com/odoo/odoo/blob/17.0/addons/pos_adyen/controllers/main.py#L15)
- [payment_stripe](https://github.com/odoo/odoo/blob/17.0/addons/payment_stripe/controllers/main.py#L70)
- [pos_viva_wallet](https://github.com/odoo/odoo/blob/17.0/addons/pos_viva_wallet/controllers/main.py#L11)

Regarding, the saas-17.1 bus communication, i am not aware about the `notify` but for the bus communication i have just used identical code like `pos_adyen`

Also, i am not fully aware about forward port. I have code ready for the saas-17.1, should i create another PR targeting saas-17.1 or fw-bot will create PR?

qle-odoo commented:
Hello 

I am sorry to inform you that Odoo will not integrate Mollie as a new payment provider in the POS.

Feel free to ask me more details by email.

Thank you for your involvement and your follow-up.

cc @vlst-odoo @caburj @avw-odoo
